homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,270 @@
+
+    @property
+    def supported_features(self) -> ValveEntityFeature:
+        """Flag supported features."""
+        return self._attr_supported_features
+
+    @property
+    def is_opening(self) -> bool | None:
+        """Return if the valve is opening or not."""
+        return self._attr_is_opening
+
+    @property
+    def is_closing(self) -> bool | None:
+        """Return if the valve is closing or not."""
+        return self._attr_is_closing
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return if the valve is closed or not."""
+        if hasattr(self, "_attr_is_closed"):
+            return bool(self._attr_is_closed)
+        if self.current_valve_position is None:
+            return None
+        return self.current_valve_position == 0
+
+    def open_valve(self) -> None:
+        """Open the valve."""
+        raise NotImplementedError()
+
+    async def async_open_valve(self) -> None:
+        """Open the valve."""
+        await self.hass.async_add_executor_job(ft.partial(self.open_valve))
+
+    def close_valve(self) -> None:
+        """Close valve."""
+        raise NotImplementedError()
+
+    async def async_close_valve(self) -> None:
+        """Close valve."""
+        await self.hass.async_add_executor_job(ft.partial(self.close_valve))
+
+    def toggle(self) -> None:
+        """Toggle the entity."""
+        fns = {
+            "open": self.open_valve,
+            "close": self.close_valve,
+            "stop": self.stop_valve,
+        }
+        function = self._get_toggle_function(fns)
+        function()
+
+    async def async_toggle(self) -> None:
+        """Toggle the entity."""
+        fns = {
+            "open": self.async_open_valve,
+            "close": self.async_close_valve,
+            "stop": self.async_stop_valve,
+        }
+        function = self._get_toggle_function(fns)
+        await function()
+
+    def set_valve_position(self, position: int) -> None:
+        """Move the valve to a specific position."""
+
+    async def async_set_valve_position(self, position: int) -> None:
+        """Move the valve to a specific position."""
+        await self.hass.async_add_executor_job(
+            ft.partial(self.set_valve_position, position)
+        )
+
+    def stop_valve(self) -> None:
+        """Stop the valve."""
+
+    async def async_stop_valve(self) -> None:
+        """Stop the valve."""
+        await self.hass.async_add_executor_job(ft.partial(self.stop_valve))
+
+    def _get_toggle_function(
+        self, fns: dict[str, Callable[_P, _R]]
+    ) -> Callable[_P, _R]:
+        if ValveEntityFeature.STOP | self.supported_features and (
+            self.is_closing or self.is_opening
+        ):
+            return fns["stop"]
+        if self.is_closed:
+            return fns["open"]
+        if self._valve_is_last_toggle_direction_open:
+            return fns["close"]
+        return fns["open"]
```

Comment:
Please test this case too.

Reply:
Done. 100% test coverage.

Reply:
Why doesn't codecov think so?

Reply:
I had 100%, but I had to do all kind of shenanigans with the internal state to hit this line on purpose. After I refactored this to exclusively use the public api (service calls) I can't find a way of hitting this line, it's logically unreachable as far as I can tell.
I can't delete it because of the type signature of the method, but it does no harm as a last safeguard.

Reply:
Doesn't this represent a case where we toggle a valve that doesn't support stop and is closing but isn't closed yet? Ie in that case we want the valve to open. That shouldn't be logically unreachable.

Reply:
100% code coverage reached